### PR TITLE
Performance enhancement with DijkstraBidirectional

### PR DESCRIPTION
**Pull request by burqen**
... and fixed ordinary Dijkstra to behave correctly.

TL;DR:
Dijkstra now works correctly in all cases but should only be used to find a specified number of shortest paths, that is the top k shortest paths. When searching for AllShortest or Single DijkstraBidirectional should be used instead.

Changes:
- Dijkstra had problems with graph that used floating point and/or zero weights.
  - Solved by introducing a tolerance level, "epsilon".
  - Needed to use STATE for this, therefore removed option for user to use STATE with Dijkstra. (This option is still available through deprecated methods in GraphAlgoFactory and Dijkstra constructors, although new and improved execution will be lost when using this.)
  - Added seperate test for Dijkstra when using STATE.
- Dijkstra didn't terminate as soon as it could.
  - Solved by introducting DijkstraEvaluator and DijkstraPathExpander. This is where STATE is occupied and therefor can not be used by user.
- Added DijkstraBidirectional.
  - Exposed through GraphAlgoFactory
  - Moved Mutable\* from ShortestPath to util as they are needed for DijkstraBidirectional
- Added k shortest paths functionality
  - This should replace the previous option of finding ALL paths which was used to find an arbitrary number of paths in increading order (seen to weight).
  - For this Dijkstra (NOT DijkstraBidirectional) is used.
  - Functionality exposed through GraphAlgoFactory
- Exchanged nested class PathInterest in BestFirstSelectorFactory
  - Now using an interface PathInterest that also provides a comparator and more options on how to determine interest of a path.
- Added Precision class for comparing doubles with range tolerance.
  - Instead of depending on apache.commons.math3
- Added ascii art for Dijkstra tests.
  - Very pedagogical
- Fixed TraversalAStar
  - Changes in BestFirstSelectorFactory changed behaviour of TraversalAStar.
- Combined Dijkstra tests into one class.

What options are available on what paths to find?:
- Dijkstra (behaviour is controlled with PathInterest)
  - All, use `PathInterestFactory.all( epsilon )` (This option should not be available as it is by far to inefficient. It is left for to not break current behavior.)
  - AllShortest, (default) use `PathInterestFactory.allShortest( epsilon )`
  - KShortest, use `PathInterestFactory.numberOfShortest( epsilon, numberOfPathsToFind )`
  - Single, just call `.findSinglePath(...)`
- DijkstraBidirectional
  - AllShortest, call `.findAllPaths(...)`
  - Single, call `.findSinglePath(...)`

Edit after review comments:
- Removed constants from Mutable*. Added links to deprecated methods.
- Made Dijkstra test parameterized. Test for fetching paths with increasing weight in seperate test class.
- Added ZERO InitialBranchState as static constant.
- Changed names from Precision to NoneStrictMath
- Added test for TopFetchingWeightedPathIterator. Added possibility to create Path in SimpleGraphBuilder and assert on path in Neo4jAlgoTestCase
- Added test for BestFirstSelectorFactory

### CODE REVIEW

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/Dijkstra.java [hunk 1]
```diff
@@ -74,18 +76,33 @@ public Dijkstra( RelationshipExpander expander, CostEvaluator<Double> costEvalua
         this( toPathExpander( expander ), costEvaluator, true );
     }
 
-    public Dijkstra( PathExpander expander, CostEvaluator<Double> costEvaluator, boolean stopAfterLowestCost )
+    public Dijkstra( PathExpander expander, CostEvaluator<Double> costEvaluator,
+            boolean stopAfterLowestCost )
     {
-        this( expander, InitialBranchState.NO_STATE, costEvaluator, stopAfterLowestCost );
+        this( expander, costEvaluator, 1.0E-8, stopAfterLowestCost );
```
Comment (MishaDemianenko) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/Dijkstra.java]: Lets introduce constant for `1.0E-8`

File: community/kernel/src/main/java/org/neo4j/graphdb/PathExpanders.java [hunk 2]
```diff
@@ -121,4 +124,70 @@ private PathExpanders()
...
+            final Paths.PathDescriptor descriptor )
+    {
+        return new PathExpander<STATE>()
+        {
+            @Override
+            public Iterable<Relationship> expand( Path path, BranchState state )
+            {
+                if ( pred.apply( path, state ) )
+                {
+                    System.out.println( Paths.pathToString( path, descriptor ) );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/graphdb/PathExpanders.java]: Can we instead provide stream as param?

**Comment by MishaDemianenko** [comment]
Can we consider to use commons-lang's Mutable\* classes and commons-math's evaluations with precision?

**Comment by burqen** [comment]
After talking to @tinwelint we decided to not pull in more dependencies than necessary. And as we only would use a very small part of those libraries it is better to just write the functionality that we need ourselves.

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/util/PathInterestFactory.java [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+import org.neo4j.function.BiFunction;
+import org.neo4j.kernel.impl.util.Precision;
+
+/**
+ * @author Anton Persson
+ */
+public class PathInterestFactory
+{
+    public static <P extends Comparable<P>> PathInterest<P> singlePathInterest()
```
Comment (tinwelint) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/util/PathInterestFactory.java]: It seems unnecessary to instantiate these PathInterest objects all the time. There should be anonymous generic-neutral private final instances inside this factory and just merely handed out via these generic static accessor methods.

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/Dijkstra.java [hunk 4]
```diff
@@ -51,74 +58,159 @@
...
+    private final boolean stateInUse;
+    // TODO: Remove stateInUse when removing deprecated constructors that uses InitialBranchState.
+    // TODO: ALso set traverser to always use DijkstaPathExpander and DijkstraEvaluator.
 
-    public Dijkstra( PathExpander expander, CostEvaluator<Double> costEvaluator )
+    /**
+     * @deprecated Dijkstra should not be used with state
+     * @param expander
+     * @param stateFactory
+     * @param costEvaluator
```
Comment (davidegrohmann) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/Dijkstra.java]: remove `@param`s there are not useful without a description

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/GraphAlgoFactory.java [hunk 5]
```diff
@@ -354,6 +360,51 @@
...
+     */
+    public static PathFinder<WeightedPath> dijkstra( PathExpander expander,
+            CostEvaluator<Double> costEvaluator, int numberOfWantedPaths )
+    {
+        return new Dijkstra( expander, costEvaluator, Precision.EPSILON,
+                PathInterestFactory.numberOfShortest( Precision.EPSILON, numberOfWantedPaths ) );
+    }
+
+    /**
+     * @deprecated Dijkstra should not be used with state on {@link PathExpander}
```
Comment (davidegrohmann) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/GraphAlgoFactory.java]: when you deprecate it would be nice to add a link to what you are suppose to use instead

File: community/kernel/src/main/java/org/neo4j/kernel/impl/util/MutableDouble.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.util;
+
+public class MutableDouble
+{
+    private static final double NULL = -1;
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/util/MutableDouble.java]: to be removed?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/util/MutableInteger.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.util;
+
+public class MutableInteger
+{
+    public static final int NULL = -1;
```
Comment (davidegrohmann) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/util/MutableInteger.java]: I have the impression that this either is not needed or is defined in the wrong place.

**Comment by davidegrohmann** [comment]
Would it make sense to add some unit tests?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/util/MutableDouble.java (hunk 6 above)
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/util/MutableDouble.java]: :+1:

File: community/kernel/src/main/java/org/neo4j/kernel/impl/util/MutableInteger.java (hunk 7 above)
Comment (burqen) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/util/MutableInteger.java]: It is only used by ShortestPath, I move it there.

**Comment by burqen** [comment]
It always make sense to test. On which part do you mean? `DijkstraBidirectional` is tested quite heavily.

**Comment by davidegrohmann** [comment]
Perhaps `BestFirstSelectorFactory`, `Paths` and `DijkstraBidirectionalPathIterator`?

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/Dijkstra.java [hunk 8]
```diff
@@ -51,74 +58,159 @@
...
+            dijkstraEvaluator = Evaluators.includeWhereEndNodeIs( end );
+        }
+        else
+        {
+            MutableDouble shortestSoFar = new MutableDouble( Double.MAX_VALUE );
+            dijkstraExpander = new DijkstraPathExpander( expander, shortestSoFar, epsilon,
+                    interest.stopAfterLowestCost() );
+            dijkstraEvaluator = new DijkstraEvaluator( shortestSoFar, end, costEvaluator );
+        }
+        return (lastTraverser = new MonoDirectionalTraversalDescription( )
```
Comment (MishaDemianenko) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/Dijkstra.java:208]: can we split assignment and return operation here to make it more readable?

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/Dijkstra.java [hunk 9]
```diff
@@ -51,74 +58,159 @@
...
+    }
+
+    public Dijkstra( PathExpander expander, CostEvaluator<Double> costEvaluator, double epsilon,
+            PathInterest<Double> interest )
+    {
+        this.expander = expander;
+        this.costEvaluator = costEvaluator;
+        this.epsilon = epsilon;
+        this.interest = interest;
+        this.stateFactory = new InitialBranchState<Double>()
```
Comment (MishaDemianenko) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/Dijkstra.java]: Same `InitialBranchState` used at https://github.com/neo4j/neo4j/pull/4879/files#diff-8a50a6e422c7dbd65844eed8a06cc7c1R89
Does it make sense to add something similar to `InitialBranchState.NO_STATE` fro those cases?

**Comment by burqen** [comment]
ping @MishaDemianenko

**Comment by MishaDemianenko** [comment]
pong @burqen :),  sorry, remember but busy with checkpointing investigation will do after that

**Comment by burqen** [comment]
Yeah, no probs. Was more of a "PR is green again"-ping

**Comment by burqen** [comment]
Gonna let @akollegger look at it first as it introduces some changes.

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/Dijkstra.java [hunk 10]
```diff
@@ -127,39 +193,67 @@ public TraversalMetadata metadata()
         return lastTraverser.metadata();
     }
 
-    private static class SelectorFactory extends BestFirstSelectorFactory<Double, Double>
+    protected static class DijkstraPathExpander implements PathExpander<Double>
```
Comment (MishaDemianenko) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/Dijkstra.java]: Why it's _protected static_ ?

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/Dijkstra.java [hunk 11]
```diff
@@ -127,39 +193,67 @@ public TraversalMetadata metadata()
...
-                Double currentAggregatedValue, Double value )
+        public PathExpander<Double> reverse()
         {
-            return withDefault( currentAggregatedValue, 0d ) + withDefault( value, 0d );
+            return new DijkstraPathExpander( source.reverse(), shortestSoFar, epsilon, stopAfterLowestCost );
         }
+    }
 
-        private <T> T withDefault( T valueOrNull, T valueIfNull )
+    protected static class DijkstraEvaluator extends PathEvaluator.Adapter<Double>
```
Comment (MishaDemianenko) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/Dijkstra.java]: here as well

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/DijkstraBidirectional.java [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+import static org.neo4j.graphdb.Direction.OUTGOING;
+import static org.neo4j.helpers.collection.IteratorUtil.firstOrNull;
+import static org.neo4j.kernel.StandardExpander.toPathExpander;
+import static org.neo4j.kernel.Traversal.bidirectionalTraversal;
+import static org.neo4j.kernel.Traversal.traversal;
+
+/**
+ * @author Anton Persson
+ */
+public class DijkstraBidirectional implements PathFinder<WeightedPath>
```
Comment (MishaDemianenko) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/DijkstraBidirectional.java]: This class needs javadocs

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/DijkstraBidirectional.java [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+                {
+                    @Override
+                    public BranchCollisionDetector create( Evaluator evaluator, Predicate<Path> pathPredicate )
+                    {
+                        return new DijkstraBranchCollisionDetector( evaluator, costEvaluator, shortestSoFar, epsilon,
+                                pathPredicate );
+                    }
+                } );
+
+        return ( lastTraverser = traversal.traverse( start, end ) );
```
Comment (MishaDemianenko) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/path/DijkstraBidirectional.java]: Can we split assigning evaluation result and return statement?

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/util/DijkstraBranchCollisionDetector.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+import org.neo4j.graphdb.traversal.Evaluator;
+import org.neo4j.graphdb.traversal.TraversalBranch;
+import org.neo4j.kernel.StandardBranchCollisionDetector;
+import org.neo4j.kernel.impl.util.MutableDouble;
+import org.neo4j.kernel.impl.util.NoneStrictMath;
+
+/**
+ * @author Anton Persson
+ */
+public class DijkstraBranchCollisionDetector extends StandardBranchCollisionDetector
```
Comment (MishaDemianenko) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/util/DijkstraBranchCollisionDetector.java:34]: Maybe it's better to move javadocs from `includePath` to class level. those will be visible to people without going to source code.

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/util/IntegerEvaluator.java [hunk 15]
```diff
@@ -35,6 +35,14 @@ public IntegerEvaluator( String costPropertyName)
 
     public Integer getCost( Relationship relationship, Direction direction )
     {
-        return ( (Number) relationship.getProperty(costPropertyName) ).intValue();
+        Object costProp = relationship.getProperty(costPropertyName);
```
Comment (MishaDemianenko) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/util/IntegerEvaluator.java]: formatting

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/util/PathInterestFactory.java [hunk 16]
```diff
@@ -0,0 +1,349 @@
...
+        private BiFunction<Double,Double,Boolean> interestFunction =
+                new BiFunction<Double,Double,Boolean>()
+                {
+                    @Override
+                    public Boolean apply( Double newValue, Double oldValue )
+                    {
+                        return NoneStrictMath.compare( newValue, oldValue, epsilon ) <= 0;
+                    }
+                };
+        private Comparator<Double> comparator = new Comparator<Double>()
```
Comment (MishaDemianenko) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/util/PathInterestFactory.java]: Same comparator used by lots of places. Does it make sense to define common in `NoneStrictMath`?

**Comment by MishaDemianenko** [comment]
@burqen few comments, otherwise looks good

File: community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/util/DijkstraBranchCollisionDetector.java (hunk 14 above)
Comment (burqen) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/util/DijkstraBranchCollisionDetector.java:34]: That comment was made more for developers who revisits this place. Because it looks like an error on first sight that branches are not pruned.
Don't think it would be of any value to user. Or do you disagree?

Comment (MishaDemianenko) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/util/DijkstraBranchCollisionDetector.java:34]: I think we need to add a note about pruning paths with 0 weights - because that can be a nasty surprise :)

Comment (MishaDemianenko) [inline community/graph-algo/src/main/java/org/neo4j/graphalgo/impl/util/DijkstraBranchCollisionDetector.java:34]: Talk and agree that its fine to have description in a method.

**Comment by akollegger** [comment]
@MishaDemianenko Great comments! For the javadoc, keep in mind the two different audiences, where public-facing classes have a how-to-use-this focus, while `*.impl.*` classes can benefit from why-does-this-exist-a perspective for implementors. 

@burqen I'm +1 on merging this when you're happy with it. Nice work!
